ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java

```diff
@@ -72,15 +96,98 @@ protected Map<String, String> doScaleOut(String startUpArgs) {
-    String jobManagerMemory = PropertyUtil.checkAndGetProperty(resource.getProperties(), JOB_MANAGER_MEMORY_PROPERTY);
+    Map<String, String> properties = resource.getProperties();
+    Map<String, String> flinkConfig = loadFlinkConfig();
+    Preconditions.checkState(properties != null && !flinkConfig.isEmpty(),
+        "resource properties is null or load flink-conf.yaml failed");
+    Long jobManagerMemory = getMemorySizeValue(properties, flinkConfig, JOB_MANAGER_MEMORY_PROPERTY,
+        JOB_MANAGER_TOTAL_PROCESS_MEMORY);
+    Long taskManagerMemory = getMemorySizeValue(properties, flinkConfig, TASK_MANAGER_MEMORY_PROPERTY,
+        TASK_MANAGER_TOTAL_PROCESS_MEMORY);
     String jobPath = getAMSHome() + "/plugin/optimize/OptimizeJob.jar";
-    long memory = Long.parseLong(jobManagerMemory) + Long.parseLong(taskManagerMemory) * resource.getThreadCount();
-    return String.format("%s/bin/flink run -m yarn-cluster -ytm %s -yjm %s -c %s %s -m %s %s",
-        getFlinkHome(), taskManagerMemory, jobManagerMemory,
-        FlinkOptimizer.class.getName(), jobPath, memory,
-        super.buildOptimizerStartupArgsString(resource));
+    long memory = jobManagerMemory + taskManagerMemory * resource.getThreadCount();
+    return String.format("%s/bin/flink run -m yarn-cluster -ytm %s -yjm %s %s -c %s %s -m %s %s",
+          getFlinkHome(), taskManagerMemory, jobManagerMemory, buildFlinkArgs(properties),
+          FlinkOptimizer.class.getName(), jobPath, memory,
+          super.buildOptimizerStartupArgsString(resource));
+  }
+
+  private Map<String, String> loadFlinkConfig() {
+    try {
+      return new Yaml().load(Files.newInputStream(Paths.get(getFlinkHome() + FLINK_CONFIG_PATH)));
+    } catch (IOException e) {
+      LOG.error("load flink conf yaml failed: {}", e.getMessage());
+      return Collections.emptyMap();
+    }
+  }
+
+  /**
+   * get jobManager and taskManager memory.
+   * An example of using Jobmanager memory parameters is as follows:
+   *    jobmanager.memory: 1024
+   *    flink-conf.jobmanager.memory.process.size: 1024M
+   *    flink-conf.yaml
+   * Prioritize from high to low.
+   *
+   */
+  protected long getMemorySizeValue(Map<String, String> properties, Map<String, String> flinkConfig,
+                                  String propertyKey, String finkConfigKey) {
+    return Optional.ofNullable(properties.get(propertyKey))
+        .map(Long::valueOf)
+        .orElseGet(() -> Optional.ofNullable(properties.get(FLINK_PARAMETER_PREFIX + finkConfigKey))
+            .map((v) -> parseMemorySize(v))
+            .orElseGet(() -> Optional.ofNullable(flinkConfig.get(finkConfigKey))
+                .map((v) -> parseMemorySize(v))
+                .orElse(0L)));
+  }
+
+  /**
+   * build user configured flink native parameters
+   */
+  protected String buildFlinkArgs(Map<String, String> properties) {
+    final Set<String> keysToFilter = new HashSet<>(Arrays.asList(
+        JOB_MANAGER_MEMORY_PROPERTY,
```

Comment:
It seems `JOB_MANAGER_MEMORY_PROPERTY` and `TASK_MANAGER_MEMORY_PROPERTY` are no longer needed to be filtered.

And should we use a static final variable to store the filter set?

Reply:
There is no need for `JOB_MANAGER_MEMORY_PROPERTY` and `TASK_MANAGER_MEMORY_PROPERTY ` here, which is my negligence. But I don't think it's necessary to use static final variables for storage, it feels meaningless.

Reply:
> It seems `JOB_MANAGER_MEMORY_PROPERTY` and `TASK_MANAGER_MEMORY_PROPERTY` are no longer needed to be filtered.
> 
> And should we use a static final variable to store the filter set?

There is no need for `JOB_MANAGER_MEMORY_PROPERTY` and `TASK_MANAGER_MEMORY_PROPERTY`  here, which is my negligence. But I don't think it's necessary to use static final variables for storage, it feels meaningless.

Reply:
My original idea was to add a static final member variable in the Class to store the filter set, in order to avoid constructing a temporary variable every time it is called. However, since this method itself is called frequently, keeping the current implementation is also okay. What do you think?

Reply:
> My original idea was to add a static final member variable in the Class to store the filter set, in order to avoid constructing a temporary variable every time it is called. However, since this method itself is called frequently, keeping the current implementation is also okay. What do you think?

Thank you for your explanation. I think this is a good idea. I'll modify the code.